stan/math/mix/functor/laplace_marginal_density.hpp

```diff
@@ -511,89 +512,165 @@ inline auto laplace_marginal_density_est(
+         &msgs](auto&& wolfe_status, auto&& wolfe_info, auto&& curr,
+                auto&& prev) {
+          wolfe_info.p_ = curr.a() - prev.a();
+          prev_g.noalias() = grad_fun(prev);
+          wolfe_info.init_dir_ = prev_g.dot(wolfe_info.p_);
+          // Flip direction if not ascending
+          if (wolfe_info.init_dir_ < 0) {
+            wolfe_info.p_ = -wolfe_info.p_;
+            wolfe_info.init_dir_ = -wolfe_info.init_dir_;
+          }
+          curr.theta().noalias() = covariance * curr.a();
+          curr.theta_grad() = theta_grad_f(curr.theta());
+          curr.alpha() = barzilai_borwein_step_size(
+              wolfe_info.p_, grad_fun(curr), prev_g, prev.alpha(),
+              wolfe_status.num_backtracks_, options.line_search.min_alpha,
+              options.line_search.max_alpha);
+          wolfe_status = internal::wolfe_line_search(wolfe_info, update_step,
+                                                     options.line_search, msgs);
+          debug::print("", 1, "Objective old: ", prev.obj(),
+                       "Objective new: ", curr.obj(),
+                       "Step size:      ", curr.alpha());
+          return abs(curr.obj() - prev.obj()) < options.tolerance
+                 || (!wolfe_status.success_ && curr.obj() <= prev.obj());
+        };
+  auto set_next_iter = [&options](auto&& curr, auto&& prev) {
+    prev.update(curr);
+    curr.alpha() = std::clamp(curr.alpha(), 0.0, options.line_search.max_alpha);
+  };
+  /**
+   * On the final loop if we found a better wolfe step, but we are going to
+   * exit, we want to make sure all of our return values are with the most
+   * recent wolfe step that was accepted. So we do one final loop to update
+   * our return values.
+   */
+  bool final_loop = false;
+  bool finish_update = false;
+  WolfeStatus wolfe_status;
+  // Start with safe step size
+  wolfe_status.num_backtracks_ = 99;
   if (options.solver == 1) {
     if (options.hessian_block_size == 1) {
+      //   std::cout << "Solver: 1Diag" << std::endl;
+      Eigen::VectorXd W_r(theta_size);
       for (Eigen::Index i = 0; i <= options.max_num_steps; i++) {
-        auto [theta_grad, W] = laplace_likelihood::diff(
-            ll_fun, theta, options.hessian_block_size, ll_args, msgs);
-        Eigen::VectorXd W_r(W.rows());
-        // Compute matrix square-root of W. If all elements of W are positive,
-        // do an element wise square-root. Else try a matrix square-root
-        for (Eigen::Index i = 0; i < W.rows(); i++) {
-          if (W.coeff(i, i) < 0) {
+        auto W = laplace_likelihood::diagonal_hessian(ll_fun, prev.theta(),
+                                                      ll_args, msgs);
+        for (Eigen::Index j = 0; j < W.size(); j++) {
+          if (W.coeff(j) < 0) {
             throw std::domain_error(
                 "laplace_marginal_density: Hessian matrix is not positive "
```

Comment:
It might be worth clarifying this error message, since it might not be exactly which Hessian matrix is being referred to: we need the negative Hessian of the log likelihood to be positive-definite